jwst/outlier_detection/utils.py

```diff
@@ -43,152 +106,185 @@ def create_median(resampled_models, maskpt, buffer_size=10.0):
+        The buffer size for the median computation, units of bytes.
 
     Returns
     -------
     median_image : ndarray
         The median image.
     """
     on_disk = resampled_models._on_disk
+    if (on_disk) and (buffer_size is None):
+        raise ValueError("Buffer size for median calculation must be provided when models are on disk")
     
-    # Compute the weight threshold for each input model
-    weight_thresholds = []
+    # Compute weight threshold for each input model and set NaN in data where weight is below threshold
     model_list = []
     with resampled_models:
         for resampled in resampled_models:
             weight_threshold = compute_weight_threshold(resampled.wht, maskpt)
+            resampled.data[resampled.wht < weight_threshold] = np.nan
             if not on_disk:
-                # handle weights right away for in-memory case
-                data = resampled.data.copy()
-                data[resampled.wht < weight_threshold] = np.nan
-                model_list.append(data)
-                del data
+                model_list.append(resampled.data)
+                resampled_models.shelve(resampled, modify=False)
             else:
-                weight_thresholds.append(weight_threshold)
-            resampled_models.shelve(resampled, modify=False)
+                resampled_models.shelve(resampled, modify=True)
         del resampled
     
-    # easier case: all models in library can be loaded into memory at once
     if not on_disk:
+        # easier case: all models in library can be loaded into memory at once
         with warnings.catch_warnings():
             warnings.filterwarnings(action="ignore",
                                     message="All-NaN slice encountered",
                                     category=RuntimeWarning)
             return np.nanmedian(np.array(model_list), axis=0)
     else:
         # set up buffered access to all input models
-        with resampled_models:
-            example_model = resampled_models.borrow(0)
-            shp = example_model.data.shape
-            dtype = example_model.data.dtype
-            nsections, section_nrows = _compute_buffer_indices(example_model, buffer_size)
-            resampled_models.shelve(example_model, modify=False)
-            del example_model
-
         # get spatial sections of library and compute timewise median, one by one
-        resampled_sections = _get_sections(resampled_models, nsections, section_nrows, shp[0])
-        median_image_empty = np.empty(shp, dtype) * np.nan
-        return _create_median(resampled_sections, resampled_models, weight_thresholds, median_image_empty)
+        resampled_sections, row_indices = _write_sections(resampled_models, buffer_size)
+        return _create_median(resampled_sections, row_indices)
 
 
-def _get_sections(library, nsections, section_nrows, imrows):
-    """Iterator to return sections from a ModelLibrary.
+def _write_sections(library, buffer_size):
```

Comment:
I think we will want this to be incremental (soon). Do you think that's within the scope of this PR? I'm thinking something like an `OnDiskMedian` class with `OnDiskMedian.add_array` and `OnDiskMedian.compute` etc. That way when we use resample to produce one resampled group at a time we can use the class to compute the median. For this PR we can just used the existing `drizzled_models` (`library` here) with the incremental class.

Reply:
Let me try to rephrase this so I can be sure I understand what you mean.

I think I might be missing a basic understanding of what a group is and does.  Is a group a list of image models that corresponds to one patch of sky, and does not overlap with patches of sky in other groups?  Meaning that it's safe to compute the median group-by-group because they don't overlap spatially?

Just trying to understand what could be changed in the code, I think you're asking for a way to easily write the sections of a single model in the library, or an arbitrary number of models, so that the median could be computed on, say, half the models.  Is that right?

Reply:
A group is a set of image models all acquired at the same time. They can be resampled together (so many input images all belonging to the same group will produce 1 grouped image). Associations will often contain multiple groups (and outlier detection only really makes sense if you have >1 and ideally >>1 group). So let's say you have an association with 20 models split into 5 groups (4 models per group). During outlier detection the call to resample many_to_many will take 20 models in, they will be grouped and 5 resampled images will be produced. These may overlap by a lot or a little (it depends on the association). The median is them computed across these 5 resampled images.
I'm not proposing any changes to the algorithm, just the way things are shuffled. I'll try to describe it here but can go into more details next week as needed. The general flow would be (within outlier detection):

- take the first group (4 images), pass them to resample to generated a single resampled image
- take the resampled image, split this into sections writing each to a temporary file
- repeat this for all groups
- when all groups are done, read in each temporary file and compute it's median (giving the median of the section taken across all resampled images) then combine these sections to get the final median image

The repeated calls to resample would be easier if the median calculation (temporary file appending) was "incremental" (able to be updated one resampled image at a time rather than requiring a list or library of resampled models).

Reply:
Thanks very much for explaining that.  Do you mean literally acquired at the same time, ie., e.g., from the two detectors of NIRCam?

I do think this refactor would be nice for reducing memory usage.  I'm not sure whether it needs to be in this PR in order for this PR to be helpful, so IMO it sort of depends on whether we feel it can get into the upcoming build or now.  If not, I'll happily take my time and refactor away.  If so, I'd say that since I'm currently working within the pre-existing code structure, I'm at least not introducing anything that makes the refactor _more_ challenging

Reply:
We decided during our stand-up that this isn't going into the build, so I'll go ahead and make this refactor too

Reply:
Does [this class](https://github.com/emolter/jwst/blob/1fdfee4deada2664110351794d40277e13941ce8/jwst/outlier_detection/utils.py#L156) look similar to what you were envisioning?

Reply:
Overall looks good. I'll give it a more detailed look during the review.

Reply:
The jwst docs have some information on grouping:
https://jwst-pipeline.readthedocs.io/en/latest/jwst/tweakreg/README.html#grouping
Although the default grouping is "all things acquired at the same time" it is user-controllable so simultaneous readouts from different detectors might be in different groups. I think that's an argument for using "grouped" instead of "timewise" in documentation/comments/names. For `OnDiskTimewiseOperation` how about just `OnDiskMedian` (since only median is supported)?

Reply:
Ok here is my current understanding of what happens during outlier detection:

- Resample_many_to_many is called, which co-adds members of the same group; this is the same statement as "co-adds different detectors of the same exposure" as in the docstring [here](https://github.com/spacetelescope/jwst/blob/be36c547939b376f9028ff449049d9507039edf9/jwst/resample/resample.py#L258). 
- The output of resample_many_to_many is a list of drizzled models.  Each member of this list is from a _different_ group, because the groups were just co-added by resample_many_to_many.  Since each drizzled frame in the list is a different group, it was taken at a different time.
- All of the drizzled models are on the same output pixel grid / WCS.  Therefore they represent, pixel-by-pixel, several representations of the same patch of sky
- The median calculation happens pixel-by-pixel, which is to say that it is finding the median brightness of every pixel over repeated exposures (i.e., over time)

While simultaneous readouts from different detectors may be in different groups, there would never be, due to physical limitations on the instrument, simultaneous readouts from different detectors that also cover the same patch of sky.  So let's say you had NIRCam A2 and A4 in different groups.  It's true they would occupy different drizzled models and would be separate in the list during median calculation, i.e. `drizzled_models = [data_a2, data_a4]`, but because they occupy different parts of the sky, you'd expect all NaNs in one of them where the other one has valid data.  For a given pixel you would have either `[NaN, valid]` or `[valid, NaN]`.  Therefore, the operation is only actually meaningful over the time axis.

I feel like we may be splitting hairs here about the name, but IMO specifying that the raison d'etre of the class is that it operates over the third axis is helpful, and in my opinion this axis can meaningfully be named the time axis per the above discussion (unless I am still misunderstanding something, which is for sure possible).

With regards to `OnDiskMedian` vs `OnDiskTimewiseOperation` or similar, 90% of the class is the same whether you're computing the median, the mean, generalized extreme-value statistics, or anything else, and I wrote a note about how this could be extended to cover those cases.  While I'm normally of the opinion that we shouldn't engineer for use-cases that don't exist yet, I anticipate a potential scenario where the name `OnDiskMedian` either becomes inappropriate or needs to be changed in the future because someone is using the class to compute a different operation on the third axis.  I'd be especially mindful of that if we move it to `stcal` and we still believe `stcal` should host observatory-agnostic stuff that might be easily usable outside our organization.  Although now that I'm writing this, maybe that is an argument for making a base class and inheriting from it...

Reply:
Thanks. I appreciate the explanation. I think `group` is more consistent with the existing API and documentation (the median is over models borrowed using `group_indices` not `time_indices` and grouped by `group_id` not `time_id`). I don't feel strongly about this so feel free to resolve the conversation if you'd like to stay with the current naming.

Reply:
I did end up changing the name to `OnDiskMedian`.  Regarding `group` vs `time`, I think the docs are relatively agnostic to the choice now.  Let me know if the changes look ok to you.

Reply:
Thanks! Marking as resolved.